docs/wslusc.1
.TH "WSLUSC" "1" "DATEPLACEHOLDER" "VERSIONPLACEHOLDER" "WSL Utilities User Manual"
.SH NAME
.B wslusc
- Component of WSL Utilities
.SH SYNOPSIS
.B wslusc
.RB [ \-IsgN ]
.RB [ \-d
.IR SHORTCUT_FILE ]
.RB [ \-e
.IR PATH ]
.RB [ \-n 
.IR NAME ]
.RB [ \-i
.IR FILE ]
.I COMMAND
.PP
.B wslusc
.RB [ \-hv ]
.SH DESCRIPTION
This is a WSL shortcut creator to create shortcut on Windows Desktop.
.SH OPTIONS
.TP
.B -h, --help
print a simple help
.TP
.B -v, --version
print current version
.TP
.B -d, --shortcut-debug \fISHORTCUT_FILE\fR
debug the generated shortcut file.
.TP
.B -g, --gui
create a shortcut that links to a GUI application. 
.TP
.B -N, --native
create a shortcut that uses the system builtin \'wslg.exe\' to launch the GUI application.
.TP
.B -i, --icon
add an icon to the Windows Shortcut. Supports \fB\.ico\fR/\fB\.png\fR/\fB\.svg\fR\.
.TP
.B -s, --smart-icon
This enables the smart icon detection. This feature requires \'wslpy\' to be installed.
.TP
.B -n, --name
add a different name other than the command name.
.TP
.B -e, --env
add a custom environment to your command. For example, if you want HiDPI support for a GTK app, add \fBexport GDK_SCALE=2;\fR in your command, like this: \fBwslusc ... \-e "export GDK_SCALE=2;" ...\fR
.TP
.B -I, --interactive
interactive mode.
.SH CONFIGURATION
.TP
.B WSLUSC_GUITYPE
Default GUI type. Default type is 'legacy'. Use 'native' for Native GUI generation (wslg.exe) by default.
.TP
.B WSLUSC_SMART_ICON_DETECTION
Smart Icon Detection. Requires wslpy (Requires Python 3.5). default is \'false\'.
.TP
.B WSLUSC_BASE_CONVERTER_ENGINE
Converter engine. You can choose either 'ffmpeg' or 'imagemagick'. By default it is 'imagemagick'.
.SH AUTHOR
Created by Patrick Wu <me@patrickwu.space>
.SH REPORTING BUGS
Report bugs to <https://github.com/wslutilities/wslu/issues>;
For Ubuntu specific/related bugs, report to <https://bugs.launchpad.net/ubuntu/+source/wslu>.
.SH COPYRIGHT
This is free software; you can redistribute it and/or modify it under
the terms of the GNU GPL version 3 or (at your option) any later
version.
There is NO warranty; not even MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
.SH SEE ALSO
wslu(7)

debian/changelog
wslu (3.2.3-0kali1) kali-dev; urgency=medium

  * wslsys: fix an issue in call_windows_uptime
    (GH: #182)
  * header: fix a distro detection issue in 
    Ubuntu Impish (21.10)
  * debian/control: fix dependency issue in 
    installing wslu (LP: #1926771)
  * docs: include available commands in manpage 
    (GH: #169)
  * project: fix a problem that Powershell output
    non-unicode character (GH: #162)
  * wslview: fix an issue that caused the relative
    Linux path unable to be opened (GH: #165)
  * wslusc: fix an issue that some people unable
    to open their GUI application (GH: #174)
  * header: Avoid error message when the file doesn't
    exist (LP: 1911379)

 -- Patrick Wu <me@patrickwu.space>  Mon, 14 Jun 2021 22:14:49 +0800

wslu (3.2.2-0kali1) kali-dev; urgency=medium

  * docs: include available commands in manpage 
    (GH: #169)
  * project: fix a problem that Powershell output
    non-unicode character (GH: #162)
  * wslview: fix an issue that caused the relative
    Linux path unable to be opened (GH: #165)
  * wslusc: fix an issue that some people unable
    to open their GUI application (GH: #174)

 -- Patrick Wu <me@patrickwu.space>  Mon, 15 Mar 2021 10:52:33 +0800

wslu (3.2.1-0kali1) kali-dev; urgency=medium

  * header: improvements on the system drive detection. (closes #133)
  * tests: update the test system and tests. (closes #125)
  * wslact: add back-compatibility for the auto-mounting feature.
  * wslview: fix an issue when handling http/https
  * wslview: fix a bug that it won't open file:///mnt/c/... or /mnt/c/...
  * wslusc: fix an issue that wsl.exe execute with wrongly passed param (close #138)
  * wslusc: fix the problem that the previous version that GUI shortcut generated won't work.
  * wslsys: fix an issue that the WSL Version shown is wrong.
  * Includes a wslview.desktop (PR #143 and #146)
  * Update the script to fix badly generated shortcut issue (fixes #141)
  * Fix a DPI script related space issue (fixes #144)

 -- Patrick Wu <me@patrickwu.space>  Wed, 23 Dec 2020 22:33:07 +0800

wslu (3.1.1-0kali1) kali-dev; urgency=medium

  * project: add separate Fedora Remix support
  * header: fix `ubuntu.exe` detection issue
  * wslview: mistake in calling a variable (fixes #118 again)
  * wslusc: always use the absolute path (fixes #129)
  * wslusc: fixes a possible white space path issue
  * wslusc: When TMP is empty, try TEMP

 -- Patrick Wu <me@patrickwu.space>  Wed, 01 Jul 2020 22:22:53 +0800

wslu (3.1.0-0kali1) kali-dev; urgency=medium

  * header: use the proper color definition
  * wslfetch: new display for WSL with WSL version displayed
  * wslfetch: use the proper logo for Ubuntu (Reported by @flexiondotorg)
  * wslsys: add WSL version option
  * wslsys: fix an issue in DPI scaling getting wrong value due to the API change in Latest fast ring insider build (Reported by @crramirez)
  * wslact: adding `--mount-options` for auto-mounting feature
  * wslact: fix an i18n related issue in time-sync feature (fixes #113 by @ricardosantos9521)
  * wslgsu: removed temporarily
  * wslgsu: limit the Windows version it can use
  * wslview: properly detect all different protocols and path styles (fixes #118)
  * package: fix alpine dependency issue (fixes #117)

 -- Patrick Wu <me@patrickwu.space>  Wed, 13 May 2020 22:53:03 +0800

wslu (3.0.1-0kali1) kali-dev; urgency=medium

  * Initial release

 -- Patrick Wu <me@patrickwu.space>  Mon, 13 Apr 2020 20:41:40 +0800


src/wslusc.sh
# shellcheck shell=bash
cname=""
iconpath=""
is_gui=0
is_interactive=0
customname=""
customenv=""
base_converter_engine=${WSLUSC_BASE_CONVERTER_ENGINE:-"imagemagick"}

help_short="wslusc [-IsgN] [-d SHORTCUT_FILE] [-e PATH] [-n NAME] [-i FILE] COMMAND\nwslusc [-hv]"

PARSED_ARGUMENTS=$(getopt -a -n "$(basename "$wslu_util_name")" -o hvd:Ie:n:i:gNs --long help,version,shortcut-debug:,interactive,path:,name:,icon:,gui,native,smart-icon -- "$@")
#shellcheck disable=SC2181
[ "$?" != "0" ] && help "$wslu_util_name" "$help_short"

function sc_debug {
	debug_echo "sc_debug: called with $*"
	dp="$(double_dash_p "$(wslvar -l Desktop)")"
	winps_exec "Import-Module 'C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\Modules\\Microsoft.PowerShell.Utility\\Microsoft.PowerShell.Utility.psd1';\$s=(New-Object -COM WScript.Shell).CreateShortcut('$dp\\$*');\$s;"
}

debug_echo "Parsed: $PARSED_ARGUMENTS"
eval set -- "$PARSED_ARGUMENTS"
while :
do
	case "$1" in
		-d|--shortcut-debug) shift; sc_debug "$@"; exit;;
		-I|--interactive) is_interactive=1;shift;; 
		-i|--icon) shift; iconpath=$1;shift;;
		-s|--smart-icon) shift; WSLUSC_SMART_ICON_DETECTION="true";shift;;
		-n|--name) shift;customname=$1;shift;;
		-e|--env) shift;customenv=$1;shift;;
		-g|--gui) is_gui=1;shift;;
		-N|--native) WSLUSC_GUITYPE="native";shift;;
		-h|--help) help "$0" "$help_short"; exit;;
		-v|--version) version; exit;;
		--) shift; cname_header="$1"; shift; cname="$*"; break;;
	esac
done
debug_echo "cname_header: $cname_header cname: $cname"
# interactive mode
if [[ $is_interactive -eq 1 ]]; then
	echo "${info} Welcome to wslu shortcut creator interactive mode."
	read -r -e -i "$cname_header" -p "${input_info} Command (Without Parameter): " input
	cname_header="${input:-$cname_header}"
	read -r -e -i "$cname" -p "${input_info} Command param: " input
	cname="${input:-$cname}"
	read -r -e -i "$customname" -p "${input_info} Shortcut name [optional, ENTER for default]: " input
	customname="${input:-$customname}"
	read -r -e -i "$is_gui" -p "${input_info} Is it a GUI application? [if yes, input 1; if no, input 0]: " input
	is_gui=$(( ${input:-$is_gui} + 0 ))
	read -r -e -i "$customenv" -p "${input_info} Pre-executed command [optional, ENTER for default]: " input
	customenv="${input:-$customenv}"
	read -r -e -i "$iconpath" -p "${input_info} Custom icon Linux path (support ico/png/xpm/svg) [optional, ENTER for default]: " input
	iconpath="${input:-$iconpath}"
fi

# supported gui check
if [ "$(wslu_get_build)" -lt 21332 ] && [[ "$gui_type" == "NATIVE" ]]; then
	error_echo "Your Windows 10 version do not support Native GUI, You need at least build 21332. Aborted" 35
fi

if [[ "$cname_header" != "" ]]; then
	up_path="$(wslvar -s USERPROFILE)"
	tpath=$(double_dash_p "$(wslvar -s TMP)") # Windows Temp, Win Double Sty.
	tpath="${tpath:-$(double_dash_p "$(wslvar -s TEMP)")}" # sometimes TMP is not set for some reason
	dpath=$(wslpath "$(wslvar -l Desktop)") # Windows Desktop, WSL Sty.
	script_location="$(wslpath "$up_path")/wslu" # Windows wslu, Linux WSL Sty.
	script_location_win="$(double_dash_p "$up_path")\\wslu" #  Windows wslu, Win Double Sty.
	distro_location_win="$(double_dash_p "$(cat ~/.config/wslu/baseexec)")" # Distro Location, Win Double Sty.

	# change param according to the exec.
	distro_param="run"

	if [[ "$distro_location_win" == *wsl\.exe* ]]; then
		if [ "$(wslu_get_build)" -ge "$BN_MAY_NINETEEN" ]; then
			distro_param="-d $WSL_DISTRO_NAME -e"
		else
			distro_param="-e"
		fi
	fi
 
	# handling the execuable part, a.k.a., cname_header
	# always absolute path
	tmp_cname_header="$(readlink -f "$cname_header")"
	if [ ! -f "$tmp_cname_header" ]; then
		cname_header="$(command -v "$cname_header")"
	else
		cname_header="$tmp_cname_header"
	fi
	unset tmp_cname_header

	[ -z "$cname_header" ] && error_echo "Bad or invalid input; Aborting" 30

	# handling no name given case
	new_cname=$(basename "$cname_header")
	# handling name given case
	if [[ "$customname" != "" ]]; then
		new_cname=$customname
	fi

	# construct full command
	#shellcheck disable=SC2001
	cname="\"$(echo "$cname_header" | sed "s| |\\\\ |g") $cname\""

	# Check default icon and runHidden.vbs
	wslu_file_check "$script_location" "wsl.ico"
	wslu_file_check "$script_location" "wsl-term.ico"
	wslu_file_check "$script_location" "wsl-gui.ico"
	wslu_file_check "$script_location" "runHidden.vbs"

	# handling icon
	if [[ "$iconpath" != "" ]] || [[ "$WSLUSC_SMART_ICON_DETECTION" == "true" ]]; then
		#handling smart icon first; always first 
		if [[ "$WSLUSC_SMART_ICON_DETECTION" == "true" ]]; then
			if wslpy_check; then
				tmp_fcname="$(basename "$cname_header")"
				iconpath="$(python3 -c "import wslpy.__internal__; print(wslpy.__internal__.find_icon(\"$tmp_fcname\"))")"
				echo "${info} Icon Detector found icon $tmp_fcname at: $iconpath"
			else
				echo "${warn} Icon Detector cannot find icon."
			fi
		fi

		# normal detection section
		icon_filename="$(basename "$iconpath")"
		ext="${iconpath##*.}"

		if [[ ! -f $iconpath ]]; then
			iconpath="$(double_dash_p "$up_path")\\wslu\\wsl.ico"
			echo "${warn} Icon not found. Reset to default icon..."
		else
			echo "${info} You choose to use custom icon: $iconpath. Processing..."
			cp "$iconpath" "$script_location"
		
			if [[ "$ext" != "ico" ]]; then
				if [[ "${base_converter_engine}" = "ffmpeg" ]] && ! type ffmpeg > /dev/null; then
					echo "The 'ffmpeg' command is needed for converting the icon with 'ffmpeg' Engine."
					if [ -x /usr/lib/command-not-found ]; then
						echo " It can be installed with:" >&2
						echo "" >&2
						/usr/lib/command-not-found ffmpeg 2>&1 | grep -E -v '(not found|^$)' >&2
					else
						echo "It can usally be installed in your package manager as 'ffmpeg'."
					fi
					exit 22
				elif ! type convert > /dev/null; then
					echo "The 'convert' command is needed for converting the icon with 'imagemagick' Engine."
					if [ -x /usr/lib/command-not-found ]; then
						echo " It can be installed with:" >&2
						echo "" >&2
						/usr/lib/command-not-found convert 2>&1 | grep -E -v '(not found|^$)' >&2
					else
						echo "It can usally be found in the imagemagick package, please install it."
					fi
					exit 22
				fi
				if [[ "${base_converter_engine}" = "ffmpeg" ]]; then
					if [[ "$ext" == "svg" ]]; then
						echo "${info} Converting $ext icon to ico..."
						echo "${warn} ffmpeg is not designed for converting svg to ico, the result may not be satisfactory."
						if ffmpeg -encoders | grep svg > /dev/null; then
							ffmpeg -hide_banner -loglevel panic -i "$script_location/$icon_filename" -width 256 -height 256 -keep_ar false -vf scale=256:256 "$script_location/${icon_filename%."$ext"}.ico"
						else
							error_echo "${warn} ffmpeg is not compiled with svg support, please compile it with svg support. Aborted." 22
						fi
						rm "$script_location/$icon_filename"
						icon_filename="${icon_filename%."$ext"}.ico"
					elif [[ "$ext" == "png" ]] || [[ "$ext" == "xpm" ]]; then
						echo "${info} Converting $ext icon to ico..."
						ffmpeg -hide_banner -loglevel panic -i "$script_location/$icon_filename" -vf scale=256:256 "$script_location/${icon_filename%."$ext"}.ico"
						rm "$script_location/$icon_filename"
						icon_filename="${icon_filename%."$ext"}.ico"
					else
						error_echo "wslusc only support creating shortcut using .png/.svg/.ico icon with ffmpeg engine. Aborted." 22
					fi
				else
					if [[ "$ext" == "svg" ]]; then
						echo "${info} Converting $ext icon to ico..."
						convert "$script_location/$icon_filename" -trim -background none -resize 256X256 -define 'icon:auto-resize=16,24,32,64,128,256'  "$script_location/${icon_filename%."$ext"}.ico"
						rm "$script_location/$icon_filename"
						icon_filename="${icon_filename%."$ext"}.ico"
					elif [[ "$ext" == "png" ]] || [[ "$ext" == "xpm" ]]; then
						echo "${info} Converting $ext icon to ico..."
						convert "$script_location/$icon_filename" -resize 256X256 "$script_location/${icon_filename%."$ext"}.ico"
						rm "$script_location/$icon_filename"
						icon_filename="${icon_filename%."$ext"}.ico"
					else
						error_echo "wslusc only support creating shortcut using .png/.svg/.xpm/.ico icon with imagemagick engine. Aborted." 22
					fi
				fi
			fi
			iconpath="$script_location_win\\$icon_filename"
		fi
	else
		if [[ "$is_gui" == "1" ]]; then
			iconpath="$(double_dash_p "$up_path")\\wslu\\wsl-gui.ico"
		else
			iconpath="$(double_dash_p "$up_path")\\wslu\\wsl-term.ico"
		fi
	fi
	
	# handling custom vairable command
	if [[ "$customenv" != "" ]]; then
		echo "${info} the following custom variable/command will be applied: $customenv"
	fi

	if [[ "$is_gui" == "1" ]]; then
		if [[ "$WSLUSC_GUITYPE" == "legacy" ]]; then
			winps_exec "Import-Module 'C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\Modules\\Microsoft.PowerShell.Utility\\Microsoft.PowerShell.Utility.psd1';\$s=(New-Object -COM WScript.Shell).CreateShortcut('$tpath\\$new_cname.lnk');\$s.TargetPath='C:\\Windows\\System32\\wscript.exe';\$s.Arguments='$script_location_win\\runHidden.vbs $distro_location_win $distro_param $customenv /usr/share/wslu/wslusc-helper.sh $cname';\$s.IconLocation='$iconpath';\$s.Save();"
		elif [[ "$WSLUSC_GUITYPE" == "native" ]]; then
					winps_exec "Import-Module 'C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\Modules\\Microsoft.PowerShell.Utility\\Microsoft.PowerShell.Utility.psd1';\$s=(New-Object -COM WScript.Shell).CreateShortcut('$tpath\\$new_cname.lnk');\$s.TargetPath='C:\\Windows\\System32\\wslg.exe';\$s.Arguments='~ -d $WSL_DISTRO_NAME $customenv $cname';\$s.IconLocation='$iconpath';\$s.Save();"
		else
			error_echo "bad GUI type, aborting" 22
		fi
	else
		winps_exec "Import-Module 'C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\Modules\\Microsoft.PowerShell.Utility\\Microsoft.PowerShell.Utility.psd1';\$s=(New-Object -COM WScript.Shell).CreateShortcut('$tpath\\$new_cname.lnk');\$s.TargetPath='$distro_location_win';\$s.Arguments='$distro_param $customenv bash -l -c $cname';\$s.IconLocation='$iconpath';\$s.Save();"
	fi
	tpath="$(wslpath "$tpath")/$new_cname.lnk"
	mv "$tpath" "$dpath"
	echo "${info} Create shortcut ${new_cname}.lnk successful"
else
	error_echo "No input, aborting" 21
fi
